### ISSUE TITLE
sci-mathematics/gap and associated packages in Prefix

### ISSUE THREAD

**Issue by strogdon** [cschwan/sage-on-gentoo#605]
In Prefix to prevent `gap-4.11.0-r1` from being pulled in when I update world I have to mask it. 
```
These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild  r  U  ] sci-mathematics/gap-4.11.0-r1 [4.10.2-r1]
[ebuild     U  ] dev-gap/SmallGrp-1.4.1 [1.3-r2]
[ebuild     U  ] dev-gap/primgrp-3.4.0 [3.3.2-r2]
[ebuild     U  ] dev-gap/GAPDoc-1.6.3 [1.6.2-r2]
[ebuild     U  ] dev-gap/transgrp-3.0 [2.0.4-r2]
[ebuild  rR    ] dev-gap/io-4.7.0
```
None of the above are in my world file. With `gap-4.11.0-r1` masked I get

```
Calculating dependencies... done!

WARNING: One or more updates/rebuilds have been skipped due to a dependency conflict:

dev-gap/GAPDoc:0

  (dev-gap/GAPDoc-1.6.3:0/0::sage-on-gentoo, ebuild scheduled for merge) USE="(prefix)" conflicts with
    ~dev-gap/GAPDoc-1.6.2 required by (sci-mathematics/gap-4.10.2-r1:0/4.10.2::sage-on-gentoo, installed) USE="(prefix) readline recommended_pkgs -emacs -vim-syntax"
    ^               ^^^^^
```
I have to mask `GAPDoc-1.6.3` to fix things. There is no issue with building Sage, just when updating world. Any ideas? Perhaps things are OK, but they seem strange to me.

**Comment by kiwifb** [comment]
gap updates are still a bit tricky. I should have sage-9.1 and master 9999 depend on gap-4.10.2 rather than 4.10.2 and above to avoid pulling them. That being said gap would show as a skipped update itself.

But for GAPDoc that would be because the dependency is on the explicit version in the gap ebuild
```
~dev-gap/GAPDoc-1.6.2
```
so of course it has to be skipped. I have started relaxing things (cross fingers) in gap-4.11.0. Got rid of those horrible slots as well. I am very much hoping that gap upgrades will behave better in the future. But the upgrade to 4.11.0 will still have bumps. I know because I did it yesterday after Volker merged the update. I just uninstalled a bunch of dev-gap packages and let the sage rebuild pull stuff in. I followed by an emerge world which upgraded quite a few dev-gap packages all by itself.

**Comment by strogdon** [comment]
I removed Sage and updated world. All the gap packages were updated. An emerge of sage now wants to pull in `sci-mathematics/sage-9.1-r3`? An explicit emerge (=sci-mathematics/sage-9999) wants to downgrade gap packages
```
Calculating dependencies... done!
[ebuild     UD ] sci-mathematics/gap-4.10.2-r1 [4.11.0-r1]
[ebuild     UD ] dev-gap/transgrp-2.0.4-r2 [3.0]
[ebuild     UD ] dev-gap/primgrp-3.3.2-r2 [3.4.0]
[ebuild     UD ] dev-gap/GAPDoc-1.6.2-r2 [1.6.3]
[ebuild     UD ] dev-gap/SmallGrp-1.3-r2 [1.4.1]
[ebuild  N    #] sci-mathematics/sage-9999  USE="debug doc-html doc-pdf jmol testsuite -X -bliss -latex" L10N="en -ca -de -es -fr -hu -it -ja -pt -ru -tr" PYTHON_TARGETS="python3_8 -python3_7"
```
which won't build because of conflicts. Removed the above and now `sage-9999` will build pulling in `gap-4.10.2-r1`. But I'm sure I'll have to mask `sci-mathematics/gap-4.11.0-r1` and `dev-gap/GAPDoc-1.6.3` to prevent world from pulling in the gap packages again.

**Comment by strogdon** [comment]
Well yesterday things built but not now
```
[knots    ] /storage/strogdon/gentoo-rap/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_8/build/lib/sage/knots/knot.py:docstring of sage.knots.knot.Knot:26: WARNING: Exception occurred in plotting knot-1
[knots    ]  from /storage/strogdon/gentoo-rap/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_8/doc/en/reference/knots/sage/knots/knot.rst:
[knots    ] Traceback (most recent call last):
[knots    ]   File "/storage/strogdon/gentoo-rap/usr/lib/python3.8/site-packages/matplotlib/sphinxext/plot_directive.py", line 472, in run_code

...

[knots    ]     sig_on()
[knots    ] sage.libs.gap.util.GAPError: Error, FreeGroup: function is not yet defined
[knots    ] /storage/strogdon/gentoo-rap/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_8/build/lib/sage/knots/knot.py:docstring of sage.knots.knot.Knot.connected_sum:25: WARNING: Exception occurred in plotting knot-2
```
Perhaps I'm jumping the gun.

**Comment by kiwifb** [comment]
gap-4.11.0 upgrade seems to be postponed to sage 9.3. :( Which version of gap gives you trouble above?

**Comment by strogdon** [comment]
The above was with `4.10.2-r1`. So maybe the update world that I performed last evening pulled in something that's not good.

**Comment by strogdon** [comment]
Yesterday my Gentoo box was updated as well and now I'm unable to build Sage on it. The reason for the failure is not obvious
```
[522/522] x86_64-pc-linux-gnu-gcc -march=native -O2 -pipe -fPIC -I/usr/lib/python3.7/site-packages/cysignals -I./sage/cpython -I/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7 -I/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/sage/ext -I/usr/include/python3.7m -I/usr/lib/python3.7/site-packages/numpy/core/include -I/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/cythonized -I/usr/include/python3.7m -c /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/cythonized/sage/ext/interpreters/wrapper_el.c -o /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/temp.linux-x86_64-3.7/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/cythonized/sage/ext/interpreters/wrapper_el.o -fno-strict-aliasing -DCYTHON_CLINE_IN_TRACEBACK=1 -std=c99
x86_64-pc-linux-gnu-gcc -shared -Wl,-O1 -Wl,--as-needed -march=native -O2 -pipe /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/temp.linux-x86_64-3.7/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/cythonized/sage/ext/interpreters/wrapper_cc.o -L/usr/lib64 -lmpfr -lmpc -lgmp -lpython3.7m -o /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/lib/sage/ext/interpreters/wrapper_cc.cpython-37m-x86_64-linux-gnu.so -lpari
x86_64-pc-linux-gnu-gcc -shared -Wl,-O1 -Wl,--as-needed -march=native -O2 -pipe /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/temp.linux-x86_64-3.7/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/cythonized/sage/crypto/boolean_function.o -L/usr/lib64 -lgmp -lpython3.7m -o /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/lib/sage/crypto/boolean_function.cpython-37m-x86_64-linux-gnu.so -lpari
x86_64-pc-linux-gnu-gcc -shared -Wl,-O1 -Wl,--as-needed -march=native -O2 -pipe /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/temp.linux-x86_64-3.7/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/cythonized/sage/ext/interpreters/wrapper_el.o -L/usr/lib64 -lpython3.7m -o /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/lib/sage/ext/interpreters/wrapper_el.cpython-37m-x86_64-linux-gnu.so -lpari
x86_64-pc-linux-gnu-gcc -shared -Wl,-O1 -Wl,--as-needed -march=native -O2 -pipe /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/temp.linux-x86_64-3.7/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/cythonized/sage/ext/interpreters/wrapper_rr.o -L/usr/lib64 -lmpfr -lgmp -lpython3.7m -o /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/lib/sage/ext/interpreters/wrapper_rr.cpython-37m-x86_64-linux-gnu.so -lpari
x86_64-pc-linux-gnu-gcc -shared -Wl,-O1 -Wl,--as-needed -march=native -O2 -pipe /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/temp.linux-x86_64-3.7/var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/cythonized/sage/calculus/riemann.o -L/usr/lib64 -lpython3.7m -o /var/tmp/portage/sci-mathematics/sage-9999/work/sage-9999/src-python3_7/build/lib/sage/calculus/riemann.cpython-37m-x86_64-linux-gnu.so -lpari
error: command 'x86_64-pc-linux-gnu-g++' failed with exit status 1
 ^[[31;01m*^[[0m ERROR: sci-mathematics/sage-9999::sage-on-gentoo failed (compile phase):
 ^[[31;01m*^[[0m   (no error message)
```

**Comment by strogdon** [comment]
I think the problem in Prefix is `dev-gap/io-4.7.0` which was upgraded from `dev-gap/io-4.6.0`.

**Comment by kiwifb** [comment]
On Gentoo, check your brial version. 1.2.9 is to be avoided in favor of either 1.2.8 or 1.2.10. The dev-gap/io is more strange unless one of its dependency needs rebuilding. I may have to look into that.

**Comment by strogdon** [comment]
My brial is OK (v. 1.2.8). It's strange since I have installed a working 9.2.beta14 for python3.7 and python3.8.

**Comment by kiwifb** [comment]
I would need the full log to backtrack that one.

**Comment by strogdon** [comment]
Ouch! I forgot this after the updates. I'm sure it's the issue.
```
# emerge -p @preserved-rebuild

Calculating dependencies... done!
[ebuild   R   ~] dev-python/cysignals-1.10.2 
[ebuild   R    ] sci-mathematics/eclib-20190909 
[ebuild   R   ~] sci-mathematics/giac-1.5.0.87 
[ebuild   R   ~] sci-mathematics/lcalc-1.23-r10 
[ebuild   R   ~] dev-python/cypari2-2.1.1 
[ebuild     U  ] www-client/chromium-85.0.4183.121 [84.0.4147.125] USE="js-type-check%* -ozone-wayland%" 
[ebuild   R   #] sci-mathematics/sage-9999
```

**Comment by kiwifb** [comment]
smells like it.

**Comment by strogdon** [comment]
`@preserved-rebuild` was the issue on Gentoo. However, I don't believe `io-4.7.0` should be used with `gap-4.10.2-r1`. Probably not an issue on Gentoo where `io-4.7.0` is unstable. But in Prefix I have to mask it. I'll close as `gap-4.11.0` is postponed.

**Comment by kiwifb** [comment]
Should work with gap-4.10+ according to the author https://github.com/gap-packages/io/blob/v4.7.0/PackageInfo.g but if you think this is an issue, I should raise the gap dependency in the in io ebuild.

**Comment by strogdon** [comment]
OK. reinstalled `io-4.7.0` and now Sage builds. When Sage wouldn't build `gap` would not start properly. Now it seems fine. Go figure. So I think the dependency is fine.
